### PR TITLE
[string performence improvement] 

### LINKED ISSUE #3259: Improve performance in functions working with String
**Summary**

Description of this feature.
Most functions working with String are not optimized well.

1. Unnecessary Extra copy inside the iteration loop.
2. Validity checks which break CPU branch prediction.

cc @mshauneu  may be interested in this.

Refer :

https://github.com/jorgecarleitao/arrow2/pull/651
https://github.com/jorgecarleitao/arrow2/pull/646

### PR DESCRIPTION
**Pull request by Veeupup**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

improve string function performance 
* almost 50% speed improvement in release mode
* 100% extra memory copy improvement in base64/quote/trim/reverse string function

before:
image

now:
image

## Changelog

- Performance Improvement

## Related Issues

Fixes #3259

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: common/functions/src/scalars/strings/quote.rs [hunk 1]
```diff
@@ -12,33 +12,37 @@
...
-                b'\'' => self.buffer.extend_from_slice(&[b'\\', b'\'']),
-                b'\"' => self.buffer.extend_from_slice(&[b'\\', b'\"']),
-                8 => self.buffer.extend_from_slice(&[b'\\', b'b']),
-                b'\n' => self.buffer.extend_from_slice(&[b'\\', b'n']),
-                b'\r' => self.buffer.extend_from_slice(&[b'\\', b'r']),
-                b'\t' => self.buffer.extend_from_slice(&[b'\\', b't']),
-                b'\\' => self.buffer.extend_from_slice(&[b'\\', b'\\']),
-                _ => self.buffer.push(*ch),
-            }
+                0 => len += buffer.write(&[b'\\', b'0']).unwrap_or(0),
```
Comment (sundy-li) [inline common/functions/src/scalars/strings/quote.rs:30]: extend_from_slice may be better， this is memory io write, no need to use Trait writer

**Comment by sundy-li** [comment]
Is the performance in debug mode or release mode?

**Comment by Veeupup** [comment]
> Is the performance in debug mode or release mode?

sorry, I test it in debug mode ... I try make it in release mode later

**Review by sundy-li — APPROVED** [review]
/LGTM

And I think it's time to add `transfrom_from_primitive` function.

**Pull request merged**